### PR TITLE
Salt specification

### PR DESCRIPTION
**Pull request by Eitu33**
Following https://github.com/massalabs/massa/pull/4044#issuecomment-1653874101

### CODE REVIEW

File: wallet/file-format.md [hunk 1]
```diff
@@ -50,6 +50,9 @@ Specifically, the PBKDF2 arguments defined in section 5.2 of the aforementioned
 - 600,000 iterations, and a
 - derived key length of 32 bytes.
 
+The salt should respect the recommended practices in [PHC string format specification](https://github.com/P-H-C/phc-string-format/blob/master/phc-sf-spec.md#function-duties), which states the following:
+> The role of salts is to achieve uniqueness. A random salt is fine for that as long as its length is sufficient; a 16-byte salt would work well (by definition, UUID are very good salts, and they encode over exactly 16 bytes). 16 bytes encode as 22 characters in B64.
+
```
Comment (Thykof) [inline wallet/file-format.md:55]: I think it doesn't bring value to copy-paste the sentence from that specification. Also we already cited https://www.ietf.org/rfc/rfc2898.txt that says:

> then the salt may be generated at random and
         need not be checked for a particular format by the party
         receiving the salt. It should be at least eight octets (64
         bits) long.

That's why I think that it's not necessary to cite PHC string format specification.

In addition, I think it's more in the Serialization of the account file that we should talk about string format: base 64 encoding; because this section is dedicated to cryptography.

In the Serialization bellow, we have to change the change the examples and replace "Salt for PBKDF2 (16 Bytes)" by something like "Salt for PBKDF2 (16 Bytes, encoded in base 64)"

**Review by Thykof — CHANGES_REQUESTED** [review]

File: wallet/file-format.md (hunk 1 above)
Comment (Eitu33) [reply wallet/file-format.md:55]: I have no issues with your suggestions, as you were part of this spec redaction I will let you apply them to avoid more friction.

**Review by litchipi — APPROVED** [review]

File: wallet/file-format.md [hunk 2]
```diff
@@ -81,7 +81,7 @@ The following table summarize the format:
 | Version | Mandatory | Integer | Entire part of this specification version | 1 |
 | Nickname | Mandatory | String || "Savings" |
 | Address | Optional | String || "AU12..." |
-| Salt | Mandatory | Byte array | Salt for PBKDF2 (16 Bytes) | [57, 125, 102, 235, 118, 62, 21, 145, 126, 197, 242, 54, 145, 50, 178, 98] |
+| Salt | Mandatory | String | Salt for PBKDF2 (16 Bytes, base 64 encoded) | "iad1ELW64ouX5HfCUCglig==" |
```
Comment (Thykof) [inline wallet/file-format.md:84]: @damip here there is the same example with trailing '=='. Should we remove it? 

also, why unpad base64? (I understand that unpadding is simple remove the trailing =, is it right?)

Comment (damip) [reply wallet/file-format.md:84]: yes, the goal is to have an unambiguous specification. Also on the rust side we need exactly that

**Review by gregLibert — CHANGES_REQUESTED** [review]

File: wallet/file-format.md [hunk 3]
```diff
@@ -107,9 +107,10 @@ In order to decrypt the private key, following steps are followed:
 
 1. User inputs a password.
 2. Password is converted to bytes (utf-8 encoding).
-3. Symmetric key is derived using PBKDF2 with this password (bytes) and the salt (bytes) as input.
-4. This derived symmetric key is then used as AES-GCM Key along with the nonce (IV) to decrypt the ciphered private key.
-5. The authentication tag at the end of ciphered data checks if the derived symmetric key used in point 4. is right.
+3. Salt is decoded from unpadded base64 to bytes.
```
[the remark this answers is not in the document]
Comment (litchipi) [reply wallet/file-format.md:110]: In the Rust library, the `Salt` struct is constructed from the base64-encoded (and unpadded) salt
However in Go, the Salt is used as a raw array of bytes.

In general, the salt is just an array of bytes, and used in the algorithms this way, the rest is just differences in the API of the libraries. So in our case, we use the base64-encoded value as-is in Rust, and decode it in Go, it depends.
(Hope that answers the question ?)

[the remark this answers is not in the document]
Comment (Eitu33) [reply wallet/file-format.md:110]: > But from my discussion with @Eitu33 I understood that we must use the base64 representation of the salt as a byte array instead.

That was the case yes but I have not followed the evolution of this matter since our last discussion, I can't confirm that this is still the current expected behaviour.

File: wallet/file-format.md [hunk 4]
```diff
@@ -46,7 +46,7 @@ To protect the private key, a symmetric key is derived from the user password us
 
 Specifically, the PBKDF2 arguments defined in section 5.2 of the aforementioned standard must follow the followings:
 
-- 16-byte salt,
+- 16-byte ,
```
Comment (Thykof) [inline wallet/file-format.md:49]: why this change?

File: wallet/file-format.md [hunk 5]
```diff
@@ -71,7 +71,7 @@ The size of the authentication tag is 16 bytes (128 bits).
 
 ### Serialization
 
-An account, in a serialized form, consists of various components, including the account's private and public keys, a salt, a nonce, a version, a nickname, and optionally, an address to aid in wallet recognition.
+An account, in a serialized form, consists of various components, including the account's private and public keys, a , a nonce, a version, a nickname, and optionally, an address to aid in account recognition.
```
Comment (Thykof) [inline wallet/file-format.md:74]: why this change?

**Review by Thykof — CHANGES_REQUESTED** [review]

File: wallet/file-format.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (gregLibert) [reply wallet/file-format.md:110]: If the content of the value is decoded before being used by the cryptography library, then why change the standard? Why don't you encode it in base64 and then decode it on the rust code side ?

**Comment by litchipi** [comment]
Closing this as this is not useful anymore: https://github.com/massalabs/massa-standards/pull/111#discussion_r1299679927

The salt spec works well now because of https://github.com/massalabs/massa/pull/4044
